File: apps/ios/Moodish/Moodish/Features/Corporate/GroupSessionViewModel.swift
import Foundation

/// Drives a group session after creation: polls every 2.5s while the
/// session is active, matching the web app's manager polling loop
/// (app.js), and exposes the state-machine-gated manager actions.
@Observable
@MainActor
final class GroupSessionViewModel {
    private let api: APIClient
    private let sessionStore: SessionStore
    let sessionId: String

    var session: GroupSession?
    var errorMessage: String?
    var isBusy = false
    private var pollTask: Task<Void, Never>?

    private var accessToken: String? {
        sessionStore.groupAccessToken(sessionId: sessionId)
    }

    init(api: APIClient, sessionStore: SessionStore, sessionId: String, initial: GroupSession? = nil) {
        self.api = api
        self.sessionStore = sessionStore
        self.sessionId = sessionId
        self.session = initial
    }

    func startPolling() {
        stopPolling()
        pollTask = Task { [weak self] in
            guard let self else { return }
            while !Task.isCancelled {
                await self.refresh()
                guard let state = self.session?.state, state.isActive else { break }
                try? await Task.sleep(nanoseconds: 2_500_000_000)
            }
        }
    }

    func stopPolling() {
        pollTask?.cancel()
        pollTask = nil
    }

    func refresh() async {
        do {
            session = try await api.getGroupSession(sessionId: sessionId, bearerToken: accessToken)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func simulateTeammates() async {
        let fixtures: [(mood: String, dietMode: String)] = [
            ("comfort food", "veg"),
            ("something light", "non_veg"),
            ("spicy and exciting", "veg")
        ]
        for (index, fixture) in fixtures.enumerated() {
            do {
                session = try await api.submitPreferences(
                    sessionId: sessionId,
                    participantId: "demo-teammate-\(index + 1)",
                    dietMode: fixture.dietMode,
                    mood: fixture.mood,
                    dietaryRules: nil,
                    allergies: nil,
                    invitePasscode: nil,
                    bearerToken: accessToken
                )
            } catch {
                errorMessage = error.localizedDescription
            }
        }
    }

    func lockAndRank() async {
        guard let accessToken else { return }
        await run { try await api.rankGroupSession(sessionId: sessionId, bearerToken: accessToken) }
    }

    func approve(optionId: String) async {
        guard let accessToken else { return }
        await run { try await api.selectGroupOption(sessionId: sessionId, optionId: optionId, bearerToken: accessToken) }
    }

    func confirmCart(addOnProductIds: [String] = []) async {
        guard let accessToken else { return }
        await run { try await api.confirmGroupCart(sessionId: sessionId, addOnProductIds: addOnProductIds, bearerToken: accessToken) }
    }

    func cancel() async {
        guard let accessToken else { return }
        await run { try await api.cancelGroupSession(sessionId: sessionId, bearerToken: accessToken) }
    }

    private func run(_ operation: () async throws -> GroupSession) async {
        isBusy = true
        errorMessage = nil
        defer { isBusy = false }
        do {
            session = try await operation()
        } catch {
            errorMessage = error.localizedDescription
        }
    }
}

File: apps/moodish_android/lib/features/corporate/group_session_view_model.dart
import 'dart:async';
import 'package:flutter/foundation.dart';

import '../../core/api_client.dart';
import '../../core/session_store.dart';
import '../../core/models/group_models.dart';

/// Polls every 2.5s while the session is active, matching the web app's
/// manager polling loop, and exposes the state-machine-gated manager actions.
class GroupSessionViewModel extends ChangeNotifier {
  final ApiClient api;
  final SessionStore sessionStore;
  final String sessionId;

  GroupSession? session;
  String? errorMessage;
  bool isBusy = false;
  Timer? _pollTimer;

  GroupSessionViewModel({required this.api, required this.sessionStore, required this.sessionId, GroupSession? initial})
      : session = initial;

  Future<String?> get _accessToken => sessionStore.groupAccessToken(sessionId);

  void startPolling() {
    _pollTimer?.cancel();
    refresh();
    _pollTimer = Timer.periodic(const Duration(milliseconds: 2500), (_) async {
      if (session != null && !session!.state.isActive) {
        _pollTimer?.cancel();
        return;
      }
      await refresh();
    });
  }

  @override
  void dispose() {
    _pollTimer?.cancel();
    super.dispose();
  }

  Future<void> refresh() async {
    try {
      session = await api.getGroupSession(sessionId: sessionId, bearerToken: await _accessToken);
      notifyListeners();
    } on ApiException catch (error) {
      errorMessage = error.message;
      notifyListeners();
    }
  }

  Future<void> simulateTeammates() async {
    final fixtures = [
      ('comfort food', 'veg'),
      ('something light', 'non_veg'),
      ('spicy and exciting', 'veg'),
    ];
    final token = await _accessToken;
    for (var i = 0; i < fixtures.length; i++) {
      try {
        session = await api.submitPreferences(
          sessionId: sessionId,
          participantId: 'demo-teammate-${i + 1}',
          dietMode: fixtures[i].$2,
          mood: fixtures[i].$1,
          bearerToken: token,
        );
      } on ApiException catch (error) {
        errorMessage = error.message;
      }
    }
    notifyListeners();
  }

  Future<void> _run(Future<GroupSession> Function(String token) operation) async {
    final token = await _accessToken;
    if (token == null) return;
    isBusy = true;
    errorMessage = null;
    notifyListeners();
    try {
      session = await operation(token);
    } on ApiException catch (error) {
      errorMessage = error.message;
    } finally {
      isBusy = false;
      notifyListeners();
    }
  }

  Future<void> lockAndRank() => _run((token) => api.rankGroupSession(sessionId: sessionId, bearerToken: token));

  Future<void> approve(String optionId) =>
      _run((token) => api.selectGroupOption(sessionId: sessionId, optionId: optionId, bearerToken: token));

  Future<void> confirmCart({List<String> addOnProductIds = const []}) =>
      _run((token) => api.confirmGroupCart(sessionId: sessionId, addOnProductIds: addOnProductIds, bearerToken: token));

  Future<void> cancel() => _run((token) => api.cancelGroupSession(sessionId: sessionId, bearerToken: token));
}
